master/internal/resourcemanagers/agent/agent_state.go

```diff
@@ -378,3 +414,274 @@ func (a *AgentState) patchSlotState(
+		Where("agent_id = ?", a.Handler.Address().Local()).
+		Scan(context.TODO())
+	if err != nil {
+		return err
+	}
+	log.Debugf("restored agent state snapshot: %v", snapshot)
+
+	return nil
+}
+
+func (a *AgentState) delete() error {
+	_, err := db.Bun().NewDelete().Model((*AgentSnapshot)(nil)).
+		Where("agent_id = ?", a.Handler.Address().Local()).
+		Exec(context.TODO())
+	return err
+}
+
+func (a *AgentState) clearUnlessRecovered(
+	recovered map[cproto.ID]aproto.ContainerReattachAck) error {
+	fmt.Println("clearUnlessRecovered", recovered)
+	updated := false
+	for d := range a.Devices {
+		if cID := a.Devices[d]; cID != nil {
+			_, ok := recovered[*cID]
+			if !ok {
+				a.Devices[d] = nil
+				a.slotStates[d.ID].containerID = nil
+				updated = true
+			}
+		}
+	}
+
+	for _, slot := range a.slotStates {
+		if slot.containerID != nil {
+			_, ok := recovered[*slot.containerID]
+			if !ok {
+				slot.containerID = nil
+				updated = true
+			}
+		}
+	}
+
+	for cid := range a.containerState {
+		_, ok := recovered[cid]
+		if !ok {
+			delete(a.containerState, cid)
+			updated = true
+		}
+	}
+
+	for cid := range a.containerAllocation {
+		_, ok := recovered[cid]
+		if !ok {
+			delete(a.containerAllocation, cid)
+			updated = true
+		}
+	}
+
+	if updated {
+		return a.persist()
+	}
+
+	return nil
+}
+
+func listResourcePoolsWithReattachEnabled() []string {
+	rpConfigList := config.GetMasterConfig().ResourcePools
+	result := make([]string, 0, len(rpConfigList))
+	for _, rpConfig := range rpConfigList {
+		if rpConfig.AgentReattachEnabled {
+			result = append(result, rpConfig.PoolName)
+		}
+	}
+
+	return result
+}
+
+// retrieveAgentStates reconstructs AgentStates from the database for all resource pools that
+// have agent_container_reattachment enabled.
+func retrieveAgentStates() (map[AgentID]AgentState, error) {
+	rpNames := listResourcePoolsWithReattachEnabled()
+
+	if len(rpNames) == 0 {
+		return map[AgentID]AgentState{}, nil
+	}
+
+	snapshots := []AgentSnapshot{}
+	err := db.Bun().NewSelect().Model(&snapshots).
+		Where("resource_pool_name IN (?)", bun.In(rpNames)).
+		Scan(context.TODO())
+	if err != nil {
+		return nil, err
+	}
+
+	result := make(map[AgentID]AgentState, len(snapshots))
+
+	for _, s := range snapshots {
+		state, err := newAgentStateFromSnapshot(s)
+		if err != nil {
+			return nil, fmt.Errorf("failed to recreate agent state %s: %w", s.AgentID, err)
+		}
+
+		result[s.AgentID] = *state
+	}
+
+	return result, nil
+}
+
+func newAgentStateFromSnapshot(as AgentSnapshot) (*AgentState, error) {
+	parsedUUID, err := uuid.Parse(as.UUID)
+	if err != nil {
+		return nil, err
+	}
+
+	slotStates := make(map[device.ID]*slot)
+	devices := make(map[device.Device]*cproto.ID)
+
+	for _, sd := range as.Slots {
+		slotStates[sd.Device.ID] = &slot{
+			device:      sd.Device,
+			containerID: sd.ContainerID,
+			enabled: slotEnabled{
+				deviceAdded:  true,
+				agentEnabled: as.UserEnabled,  // TODO
```

Comment:
what is this TODO for?

Reply:
that's the same as TODO circa line 435: it'd be useful to better disambiguate and simplify user settings (explicit disable and enable) vs "state" one (e.g. preDisconnectEnabled and preDisconnectDraining)